### PR TITLE
Feat/filters event VM and SC

### PR DESCRIPTION
**Pull request by e-bernier**
## Description

Refactors and expands event-related state management logic to improve modularity, reactivity, and test coverage.  
Integrates `MapEventStateController` with `EventViewModel` and updates corresponding unit tests for broader feature validation.

**Related Issue:** #218 #192 

---

## Changes

**Implementation:**
- Refactored `EventViewModel` to delegate event refresh and filtering to `MapEventStateController`.
- Enhanced `MapEventStateController` to handle:
  - Event filtering via `FiltersSectionViewModel`.
  - Management of joined, saved, and filtered event lists.
  - Search by title, description, tags, and location.
  - User-based event synchronization.
- Improved state refresh and error-handling mechanisms.
- Introduced cleaner dependency injection patterns for testability.

**Tests:**
- Added comprehensive tests for `MapEventStateController`.
- Extended `MapScreenViewModelTest` with coverage for event selection, saving/joining logic, and UI state transitions.
- Validated integration between filters, search, and event state updates.

---

## Testing
- [x] Unit tests pass  
- [x] Instrumentation tests pass  
- [x] Manual testing completed  
- [x] Coverage maintained (≥80%)

**Test summary:** Added unit tests for event filtering, joining/saving logic, and MapScreenViewModel integration. Coverage increased across event-related components.

---

## Screenshots/Videos

_N/A – Logic and test-focused PR._

---

## Checklist
- [x] Sufficient documentation and minimal inline comments  
- [x] All tests passing  
- [x] No new warnings  
- [x] If related issue exists: issue has all labels, fields, and milestone filled

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt [hunk 1]
```diff
@@ -1,217 +1,319 @@
...
+          emptyList()
+        } else {
+          allEvents.filter {
+            it.title.trim().lowercase().contains(lowerQuery) ||
+                it.description.trim().lowercase().contains(lowerQuery) ||
+                it.tags.any { tag -> tag.lowercase().contains(lowerQuery) } ||
+                it.location.name.lowercase().contains(lowerQuery)
+          }
+        }
+  }
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt]: **Error**: searchEvents accesses title/description/tags/location.name without null-safety. Make access null-safe or assert non-nullability in Event model. Add unit tests with events that have missing or empty fields to ensure search does not throw.

File: app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt [hunk 2]
```diff
@@ -673,23 +666,23 @@ class MapScreenViewModel(
   }
 
   fun isEventSaved(event: Event): Boolean {
-    return savedEventIds.contains(event.uid)
+    return savedEvents.contains(event)
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt]: **Nitpick**: isEventSaved uses savedEvents.contains(event) relying on Event equality. Use uid-based equality check (savedEvents.any { it.uid == event.uid }) or document Event equality semantics and add tests for different object instances representing the same event.

File: app/src/main/java/com/swent/mapin/ui/event/EventViewModel.kt [hunk 3]
```diff
@@ -2,77 +2,29 @@ package com.swent.mapin.ui.event
...
-        _error.value = e.message
-      }
-    }
-  }
-
   fun addEvent(event: Event) {
     viewModelScope.launch {
       try {
         eventRepository.addEvent(event)
+        stateController.refreshEventsList()
```
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/event/EventViewModel.kt]: **Question**: calling refreshEventsList after every mutation forces a full list refresh which may be inefficient. Consider returning the updated event from the repository or using event-level updates to avoid full list reloads. Add tests that assert refresh is called only when required and that performance is acceptable.

File: app/src/androidTest/java/com/swent/mapin/ui/map/BottomSheetContentTest.kt [hunk 4]
```diff
@@ -37,6 +40,10 @@ class BottomSheetContentTest {
   val filterViewModel = FiltersSectionViewModel()
   val locationViewModel = LocationViewModel()
   val profileViewModel = ProfileViewModel()
+  val eventStateController = mockk<MapEventStateController>()
+  val eventViewModel =
+      EventViewModel(
+          eventRepository = LocalEventRepository(), stateController = eventStateController)
```
Comment (naderbahri) [inline app/src/androidTest/java/com/swent/mapin/ui/map/BottomSheetContentTest.kt]: **Nitpick**: tests now instantiate a real EventViewModel with a LocalEventRepository in many places. Replace heavy integration-style instantiation with a mocked or fake EventRepository and a mocked MapEventStateController where behavior needs to be controlled. Add tests focused on UI → controller contract rather than full repo network calls.

File: app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt (hunk 1 above)
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt]: These variables are non nullable so shouldn't happen but change done.

File: app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt [hunk 5]
```diff
@@ -1,217 +1,319 @@
...
+   * @param query The search query string.
+   */
+  fun searchEvents(query: String) {
+    val lowerQuery = query.trim().lowercase()
+    searchResults =
+        if (lowerQuery.isEmpty()) {
+          emptyList()
+        } else {
+          allEvents.filter {
+            it.title.trim().lowercase().contains(lowerQuery) ||
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt]: This is already an improvement from the previous version that was fetching in the list from firebase again. But yes i plan to make a performance focused PR for these kind of things

File: app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt [hunk 6]
```diff
@@ -1,217 +1,319 @@
...
+            // Update user's participatingEventIds
+            val userProfile =
+                userProfileRepository.getUserProfile(currentUserId)
+                    ?: throw Exception("User profile not found")
+            if (!userProfile.participatingEventIds.contains(event.uid)) {
+              val updatedParticipatingEventIds =
+                  userProfile.participatingEventIds.toMutableList().apply { add(event.uid) }
+              val updatedUserProfile =
+                  userProfile.copy(participatingEventIds = updatedParticipatingEventIds)
+              userProfileRepository.saveUserProfile(updatedUserProfile)
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/ui/map/eventstate/MapEventStateController.kt]: Same as above

File: app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt [hunk 7]
```diff
@@ -68,17 +70,14 @@ class MapScreenViewModel(
           sheetConfig = sheetConfig,
           initialState = initialSheetState,
           isProgrammaticZoom = { cameraController.isProgrammaticZoom })
-  private val eventStateController =
+  val eventStateController =
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt]: Yes this will be done in the next PR. A lot more changes than that will be done in MapScreenViewModel.

File: app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt [hunk 8]
```diff
@@ -367,7 +360,7 @@ class MapScreenViewModel(
 
   private fun applyEvents(newEvents: List<Event>) {
     _events = newEvents
-    eventStateController.updateBaseEvents(newEvents)
+    eventStateController.refreshEventsList()
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/ui/map/MapScreenViewModel.kt]: Once again, as above.
